### ISSUE TITLE
Replace github.com/masterminds/sprig template functions

### ISSUE THREAD

**Issue by twpayne** [twpayne/chezmoi#2668]
## Is your feature request related to a problem? Please describe.

`github.com/masterminds/sprig`, used by chezmoi,  has unfortunately become a popular library of template functions despite several critical issues:

* It is unmaintained.
* The argument order of its functions is incompatible with Go template's pipelining (see #2215).
* Its naming convention does not match Go's (e.g. `toJson` should be `toJSON`).

## Describe the solution you'd like

chezmoi's template functions should be up-to-date, compatible with Go template pipelining, and match Go's naming conventions.

## Describe alternatives you've considered

Keeping minorminds/sprig template functions as-is.

## Additional context

A smooth transition should be provided, with both old and replacement template functions being available to users for some time. Users can use `.chezmoiversion` to help the transition.

**Comment by bradenhilton** [comment]
Perhaps we could also take the opportunity to implement a higher-order function (which would add to the `funcMap`, not something done at runtime) similar [to the one in this issue](https://github.com/Masterminds/sprig/issues/91)?

As you know, there is the [`quoteList`](https://www.chezmoi.io/reference/templates/functions/quoteList/) template function, which was implemented explicitly. So far, there hasn't been much of a demand sufficient to add similar functions, which would probably be necessary to add an explicit function (and thus create clutter and perhaps some complexity).

An `each` function separates the iteration logic from the action being taken on the item in the list. Going forward, implementing `stringFuncEach` would likely be as simple as implementing `stringFunc` and then setting `funcMap["stringFuncEach"]` to something like `eachFunc(stringFunc)`.

The `each{Func}` functions could be as generic as possible or type specific if needed.

Maybe this could even be done dynamically with reflection? So we could do something like:

1. Group functions by type if needed
2. Iterate through each function/group
3. Get function name (with reflection?)
4. `funcMap[funcName + "Each"] = eachFunc{Type}(func)` (probably not syntactically correct)

Separating the logic would remove the need for explicit functions in all but the most specialized cases, which would remove clutter and be trivial to extend as new functions are added. 

The alternative, which would be better, but also likely increase complexity, would be to add iteration logic to every requisite function, so that you can pass it a single item or a list of items and get the expected output, e.g.

```
$string | quote
"foobarbaz"


$list | quote
["foo", "bar", "baz"]
```

This approach *would* be easier if we had control over how each function is implemented, as sprig is essentially a black box in this case.

Likely v3 because this would mean removing/replacing/renaming `quoteList` at the very least.

Interested in any thoughts anyone might have.

**Comment by twpayne** [comment]
> Perhaps we could also take the opportunity to implement a higher-order function

The tricky thing here is that `text/template` is, by design, a simple language and does not support higher-order functions, and is unlikely to ever do so in the future. So, without permanently forking `text/template` (which would be an immeasurable burden), this is not possible to do.

**Comment by bradenhilton** [comment]
Doesn't it just add function permutations to the `funcMap`?

I know it's probably not possible to do something like

```
{{ $list | map | squote }}
```

but that is different, no?

I tried dropping the example in the linked issue in to test but it's not compatible without some tweaks (`...interface{}` vs `[]string` etc.) and I ran out of time that day so I just left it there.

Are you saying it's impossible to do something like

```go
funcMap["squoteEach"] = each(squote)
funcMap["add1Each"] = each(add1)
```

without a `text/template` fork?

Making all template functions list-aware somehow is probably more user friendly anyway, but I'm not sure exactly what that implementation would look like.

Should we create a `funcs` repo in the chezmoi organization?

**Comment by twpayne** [comment]
Ah, sorry I misunderstood. My misunderstanding was that I thought that `each` was to be a **template function** that took a function and a list as arguments and returned a new list of the function applied to each element in the list. This isn't possible with `text/template` as `text/template` does not support functions as values.

If I now I understand correctly, `each` should be a Go function that makes it easier to write template functions that apply a Go function to every element in a list. This is certainly possible, and doesn't need to wait for v3 as it's an internal implementation detail that is not visible to users (with the exception of a few more template functions being added).

I have to ask, though: given that `text/template` already supports `range`/`end` actions, how much are these extra functions really needed? `quoteList` is certainly useful as you often need to quote multiple strings for command arguments, but what other functions do you anticipate being regularly useful?

**Comment by twpayne** [comment]
Note it's also possible to extend template functions so that they operate on lists as well as isolated values. For example, we can extend `quote` to return a list of quoted strings if it is passed a list of strings, without breaking its existing functionality of returning a quoted string if passed a single string.


> Should we create a `funcs` repo in the chezmoi organization?

Hmm, yes. I'll create a `templatefuncs` repo with some initial ideas.

**Comment by bradenhilton** [comment]
> Note it's also possible to extend template functions so that they operate on lists as well as isolated values.

I think we should just do this. I'd argue that `{{ $list | quote }}` implicitly iterating through the list and quoting each element is expected behavior (as opposed to something like implicitly joining the elements, then quoting) and doesn't increase the number of template functions.

**Comment by twpayne** [comment]
So, sadly, thanks to Masterminds/sprig's inherent brokenness, this isn't as immediately fixable as I had hoped. Some specific problems:

* Masterminds/sprig's `quote` function already accepts lists ... and quotes them:
   ```console
   $ chezmoi execute-template '{{ list "a" "b" | quote }}'
   "[a b]"
   ```
   so changing `quote`'s behavior for lists will be a backwards-incompatible change.
* Masterminds/sprig just seems more and more broken the more I look at it. For example, here's [the implementation of `squote`](https://github.com/Masterminds/sprig/blob/581758eb7d96ae4d113649668fa96acc74d46e7f/strings.go#L93-L101). It's not even slightly correct: strings with a single quote in them are not correctly escaped 🤯 .

So, replacing the template library is definitely something for chezmoi v3, as trying to maintain backwards compatibility with all of Minorminds stupidity is just not possible.

**Comment by bradenhilton** [comment]
Also, just to clarify as I didn't do it before...

> I have to ask, though: given that `text/template` already supports `range`/`end` actions, how much are these extra functions really needed?

You're absolutely right, **especially if the solution was to add `*Each` permutations of each function to the `funcMap`**, but if we control the implementations of the functions this is a non-issue, as we can make them list-aware. This has some advantages over `range`/`end`:

1. Improved template readability. `range`/`end` require whitespace management. Often I want to do something like this:
    
    ```
    {{ range $list }}
        {{ . | quote }}
    {{ end }}
    ```

    for readability, but the whitespace is included in the output, so it must be removed with the control characters. This is one of the simplest examples I could come up with and it's still not easy to reason about at a glance in my opinion.

    Also, it's not like this approach would become unusable, it would still work perfectly as a backup or a preference depending on the user.

    I'm not sure if I can articulate this properly, but this approach is also entirely focused on the *output*, it doesn't affect the `$list` variable in any way, so it's difficult to pipe the result to something else. Speaking of which...

2. Improved pipelining. This removes the need for temporary variables/variable re-assignment/multi-line actions etc., which should allow something like the following:
    
    ```
    {{ $numList | add1 | quote }}  # here I am assuming that the functions work correctly with lists
    ```

**Comment by twpayne** [comment]
https://github.com/chezmoi/templatefuncs contains some initial replacement template functions. It's a work in progress.

**Comment by halostatue** [comment]
I was looking at [gomplate](https://github.com/hairyhenderson/gomplate) the other day and it seems to me that it has a few nice things about it and may be a viable replacement for sprig in a number of places; it has other features that aren't as applicable to Chezmoi (but many of them would be, I think).

The biggest thing that I liked about it is that it exposes *objects* rather than functions (although some functions are common enough to expose aliases), so for v3 (#2673) it would be possible to have `onepassword.Details` instead of `onepasswordDetails` &c.

**Comment by bradenhilton** [comment]
@halostatue I'd personally rather take inspiration from gomplate than integrate it into chezmoi.

If gomplate went the way of sprig, we would just find ourselves in this situation again. Creating our own bespoke function library allows us to fix and extend things quickly, rather than having to patch in our own fixes and create discrepancies like the one mentioned recently in https://github.com/twpayne/chezmoi/issues/3335:

> On a related note, `toPrettyJson` is [listed on chezmoi's website](https://www.chezmoi.io/reference/templates/functions/toPrettyJson/), which suggests that you're rolling your own implementation instead of just using sprig's. If that suggestion is false then I recommend removing that page.

Using objects just isn't necessary for some use cases. Would we really benefit from having something like `strings.ToUpper` instead of `toUpper`? The aliases would just create clutter in the FuncMap. You also end up with awkward names like `conv.ToString` and `coll.Dict` to avoid having the names be too verbose.

**Comment by halostatue** [comment]
For some operations, sure having both the alias and the full version would be a waste. On the other hand, collecting password manager functionality into objects with functions available would IMO be *much* cleaner.

Providing migration steps for this would be vital (e.g., reimplement the 1Password functionality as `onepassword.*` or `op.*`, but put the aliases in place; start warning about the use of those aliases in a future release, and remove them at Chezmoi 3).

The warning phase could even be done as "light" (the use of a deprecated function/alias warns once) and "heavy" (*each* use of a deprecated function/alias warns).

This sort of approach would allow us to *think* about what functions should be quickly available (`toString`, `toUpper`, `toJson`, `fromJson`, etc.) and what functions should be grouped together.

I may put together a list of the functions available in chezmoi via sprig or direct and suggest some sort of organization.

**Comment by bradenhilton** [comment]
Generally I agree, but I think we need to get https://github.com/chezmoi/templatefuncs in a much more complete state before we even think about that.

I think v3 should be a clean break, and we should be as noisy as would be reasonable about that leading up to its release (once the new features are nearing completion).

**Comment by twpayne** [comment]
So, I propose to make this change in an incremental way, without needing chezmoi v3, with the following strategy:

1. Add a `template.functions` config variable which is a list of strings in priority order.
2. The list element `github.com/masterminds/sprig` ("sprig") corresponds to the existing minorminds/sprig functions and the list element `github.com/chezmoi/templatefuncs` corresponds to https://github.com/chezmoi/templatefuncs ("templatefuncs").
3. The default value of `template.functions` will be `github.com/masterminds/sprig`.
4. Users can add templatefuncs to the end of the list, but sprig functions will take priority.
5. When the user has updated their templates to use templatefuncs they can move templatefuncs to the head of the list.
6. When the user no longer relies on any sprig functions they can remove sprig from the list.
7. In chezmoi v3 we remove the sprig option.

This should allow an orderly transition from sprig to templatefuncs, and of course the user can use `chezmoi diff` and chezmoi's error reporting to catch any problems along the way. As `template.functions` will be set in the user's configuration file, they can even use different template functions on different systems (if they are mad enough).

@bradenhilton thoughts?

**Comment by halostatue** [comment]
This seems like a reasonable approach to me.

**Comment by bradenhilton** [comment]
@twpayne How do you propose we handle the current chezmoi-specific functions, especially functions such as `ioreg` and our overridden sprig functions? My understanding was that `ioreg` would be removed, and the rest would eventually be rolled into templatefuncs when ready, but allowing templatefuncs now complicates that slightly.

I do think incremental adoption is a good idea.

**Comment by 2-4601** [comment]
Just came across this fork of sprig: https://github.com/go-sprout/sprout

Here's an issue with some discussion on roadmap: https://github.com/go-sprout/sprout/issues/1
More discussion on improvements here: https://github.com/orgs/go-sprout/discussions

I do not know if sprout aims to solve the problems raised in the first post of this issue. This is just a FYI :)

**Comment by 42atomys** [comment]
Hi everyone, as currently the main maintainer of [go-sprout/sprout](https://github.com/go-sprout/sprout), I wanted to provide some additional context and insights regarding your initial issue.

> The argument order of its functions is incompatible with Go template's pipelining

Firstly, I agree this is something we should address in order to improve compatibility.

> Its naming convention does not match Go's

Secondly, the naming convention of sprig does not match Go's, which could lead to confusion or issues when using the library.
Additionally, I'd like to highlight the [aliasing function feature](https://docs.atom.codes/sprout/function-aliases), we use internally for backward-compatibility reasons. You can learn more about this by checking out the alias.go file on our repository (https://github.com/go-sprout/sprout/blob/main/alias.go#L9-L46).

---
Giving a second life to sprig with sprout is an exciting challenge due to who use sprig (like Helm, tempo and Argo), we need to involve with backward-compatibility to don't break end-users configuration and go forward too.

This is a complex issue, but I'm confident we can find a solution that balances both needs.

If you have any questions or concerns, please don't hesitate to reach out on our discussions page (https://github.com/orgs/go-sprout/discussions) or ping me directly (@42atomys).


Have a great day everyone!

**Comment by twpayne** [comment]
Re-kindling an old issue here...

https://github.com/go-sprout/sprout looks very promising, particularly thanks to its backwards-compatibility with sprig and forward-looking functions.

chezmoi already overrides several of sprig's functions to provide better type compatibility:

https://github.com/twpayne/chezmoi/blob/ec245dd624b1e7173d4a10776ca221fccb0e84bc/internal/cmd/config.go#L462-L470

Switching to sprout looks like a better long-term option. Specific things that should happen (in no particular order):

* chezmoi should add a `template.functions` configuration variable to allow users to choose between sprig and sprout as the default template functions.
* Optionally, this set of functions should be settable per-file, maybe using a [template directive](https://www.chezmoi.io/reference/templates/directives/).
* We should check functionality with https://github.com/chezmoi/templatefuncs, and port any missing functionality to sprout.

Refs #4808.

**Comment by 42atomys** [comment]
Hi, just receive a notification from here.
Since my last intervention I’ve been wondering about the backward-compatibility impact on end users.

I know migrating away from the old library is a big task. To help developers ease the transition, I have implemented a **Functions Notices** feature (see https://docs.atom.codes/sprout/features/function-notices). With this mechanism, when a template uses a deprecated function or an old signature or an old function, your application logs a warning, which gives users time to migrate gracefully without immediate breakage.

In projects I maintain, I follow a policy of issuing deprecation warnings for about five versions before removing a feature completely. That way we avoid breaking changes, and give users even those who don’t always read the changelog, several releases (≈ 6 months) to adapt.

If you think it’s useful, I’d be happy to help and answer your questions.

**TL;DR:** A feature of Sprout can drastically help you and your end-users to migrate without fear and frustrations.

**Comment by twpayne** [comment]
Thank you very much @42atomys!

As I understand it, there are effectively three sets of template functions:

* `sprig`, the original and buggy.
* `sprigin`, which is sprout's compatibility layer with `sprig`. It's very similar to `sprig`, but with a few behavioral differences and generates nice deprecation notices.
* `sprout`, template functions done properly.

For chezmoi, I'm thinking of:

* Individual templates can choose which set of functions they use with a [template directive](https://www.chezmoi.io/reference/templates/directives/). Specifically, they'll be able to specify `chezmoi:template:functions=sprig`, `chezmoi:template:functions=sprigin`, or `chezmoi:template:functions=sprout`.
* At a global level, a `template.functions` [configuration variable](https://www.chezmoi.io/reference/configuration-file/variables/) will allow the user to set the default set of functions to use, either `sprig`, `sprigin`, or `sprout`, with `sprig` as the initial default.

This way, users can migrate on a file-per-file basis.

As I care a lot about maintaining backwards compatibility in minor versions, I would only make `sprigin` the default after a few months. Maybe sometime in the future I'll remove `sprig` completely. If chezmoi version 3 ever happens, then I would remove both `sprig` and `sprigin`.

As well as the code changes, this will need some restructuring of the documentation.

**Comment by 42atomys** [comment]
Thanks for the clear plan. Here’s how I see it in the context of the Sprout migration.

Sprout already provides a [documented migration path](https://docs.atom.codes/sprout/migration-from-sprig) from Sprig compatibility mode to full Sprout usage, with deprecation notices to help users transition smoothly and avoid breakages.

Given that:

* **`sprigin` is intentionally a seamless, non-breaking replacement for `sprig`**: it fixes panics and adds deprecation warnings while preserving existing behavior, so templates continue to work as-is.
* Switching the default from `sprig` to `sprigin` in a **minor release** feels safe: users benefit from safer behavior and warnings without being forced to change anything.
* Exposing both a per-template directive and a global `template.functions` config that let users choose between:
  * `v1 = sprigin` (compatibility + warnings)
  * `v2 = sprout` (modern functions)

  gives users real flexibility to migrate and fits very well with your template directive feature 🎉.

Introducing a three-step migration (`sprig` -> `sprigin` -> `sprout`) does not add much value for end users and may even cause confusion without context. A simple `v1` / `v2` model would be clearer and more user-friendly.

For extra confidence on the **chezmoi** side, if you have small or large test files, could you share them? I can run the tests against your existing templates and test suites before anything lands, to make sure behavior stays unchanged.

Happy to help with mappings or test validation before any merge.

**Comment by felipecrs** [comment]
As a user of chezmoi, thanks a lot for working on replacing sprig! Its design inconsistencies really hurt the UX.

**Comment by twpayne** [comment]
Thank you @felipecrs!

As you're one of the most advanced users of chezmoi and templating, would you be able to test #4826 with your configuration? This PR switches the template functions from sprig to sprout using sprout's compatibility layer, so the change should be transparent. Any differences you find will be super helpful for both chezmoi and sprout :)

**Comment by halostatue** [comment]
I'll also give it a try soon.

**Comment by felipecrs** [comment]
@twpayne, there was no difference at all running #4826 against my dotfiles. The compatibility layer is working really great. :)

**Comment by twpayne** [comment]
Awesome, thank you very much for the testing @felipecrs!

**Comment by 42atomys** [comment]
Really nice to see all are working on, thanks you @felipecrs for your tests ! And @twpayne for your reactivity

**Comment by LEI** [comment]
I am observing one difference in templates with `dig`:

```
{{ warnf "1: %v" (dig "a" "b" (dict "a" (dict "b" 1))) }}
{{ warnf "2: %v" (dig "a" "b" nil (dict "a" (dict "b" 1))) }}
```

```bash
$ chezmoi apply --dry-run
chezmoi: warning: 1: map[b:1]
chezmoi: warning: 2: 1

$ go tool chezmoi apply --dry-run
chezmoi: warning: 1: 1
chezmoi: warning: 2: <nil>
```

It seems expected, and this simplifies the syntax.

> Previously, the dig function would return the last map in the access chain

https://docs.atom.codes/sprout/migration-from-sprig#dig

**Comment by 42atomys** [comment]
> I am observing one difference in templates with `dig`:
> 
> ```
> {{ warnf "1: %v" (dig "a" "b" (dict "a" (dict "b" 1))) }}
> {{ warnf "2: %v" (dig "a" "b" nil (dict "a" (dict "b" 1))) }}
> ```
> 
> ```bash
> $ chezmoi apply --dry-run
> chezmoi: warning: 1: map[b:1]
> chezmoi: warning: 2: 1
> 
> $ go tool chezmoi apply --dry-run
> chezmoi: warning: 1: 1
> chezmoi: warning: 2: <nil>
> ```
> 
> It seems expected, and this simplifies the syntax.
> 
> > Previously, the dig function would return the last map in the access chain
> 
> https://docs.atom.codes/sprout/migration-from-sprig#dig


Thanks @LEI to raise it, the dig function don't have the signature warn and backward, this is a bug because not desired ! I will handle this issue on my side for sprigin ! I will keep you up to date with fixed version

**Comment by 42atomys** [comment]
This issue are now resolved and an explicit warning in `sprigin` are added (version: `v1.0.3`). 
Correct information added on doc : https://docs.atom.codes/sprout/migration-from-sprig#dig :) 

Enjoy your holidays everyone !

**Comment by marmitar** [comment]
I forgot to test this one early, but `derivePassword` is not working for me on v2.69.0.

```console
$ chezmoi execute-template '{{ derivePassword 1 "long" "password" "user" "example.com" }}'
<no value>
$ chezmoi execute-template '{{ derivePassword 1 "long" "password" "user" "example.com" | printf "%s\n" }}'
%!s(<nil>)
```

Edit: I could open another issue with more details, if you prefer.

**Comment by twpayne** [comment]
> I forgot to test this one early, but `derivePassword` is not working for me on v2.69.0.

Thank you for reporting this. This one is really, really weird. I've opened https://github.com/go-sprout/sprout/pull/157 with a fix that I don't understand.

**Comment by marmitar** [comment]
Found an issue with `dig` too: go-sprout/sprout#159

**Comment by twpayne** [comment]
As there were more compatibility problems between sprig and sprigin, I'll revert to sprig now in #4877 and then follow the incremental approach described in https://github.com/twpayne/chezmoi/issues/2668#issuecomment-3633347102.

**Comment by 42atomys** [comment]
To keep track of recent iteration on `sprout/sprigin`, 
In next following weeks I will release `v1.0.4` with recent bugs found from chezmoi community. With a clear vision of what changes. I will release `v1.0.4` only when we are sure no more incompatibility are found, based on work of @twpayne on https://github.com/go-sprout/sprout/pull/165. I will cover every function with fuzzy testing ! 🪨

I can suggest waiting for the release of `v1.0.4` before starting to implement the incremental approach or reimplement sprigin

**Comment by 42atomys** [comment]
Hi mates 👋

It’s been a long time! After a few months of testing `v1.0.4` without any new issues, `v1.1.0` has finally been released, with a clear separation between `sprout` and `sprigin`.

`sprigin` is the backward-compatible package, fully compatible with Sprig, with warning messages explaining what needs to be migrated before switching to `sprout`. And `sprout` no longer do magic to detect template formating.

This allows you to take the incremental approach more easier when migrating from `sprigin` to `sprout` 💜 
Don't hesitate to @ me if needed 🌱 

_Full release note: https://github.com/go-sprout/sprout/releases/tag/v1.1.0_

**Comment by marmitar** [comment]
@42atomys Just built chezmoi from d969d39, which includes v1.0.4, but `derivePassword` seems to still have issues. It's not empty anymore, but returning `���\u00010�`. Edit: seems to be the only issue, though.

**Comment by marmitar** [comment]
Similar issue to last time: the `uint32` parameter. I opened go-sprout/sprout#197 fixing it.

**Comment by 42atomys** [comment]
@marmitar thanks mate, [v1.1.1](https://github.com/go-sprout/sprout/releases/tag/v1.1.1) with your fix are pushed !

**Comment by 2-4601** [comment]
Thanks 42atomys for the sprout library, and thanks twpayne for considering it for chezmoi and laying out a conservative transition path. I noticed that twpayne recently committed some changes related to this transition, and wanted to test how my dotfiles work with the sprigin compatibility layer.

I downloaded a release artifact [chezmoi-linux-amd64](https://github.com/twpayne/chezmoi/actions/runs/33039135877/artifacts/9633288891) from commit 50b32b54ca5479935b53b408afa6b49d95996371.

While trying to run chezmoi apply with this new version, I noticed that math functions don't work when both arguments are strings.
```
$ chezmoi-50b32b54c --version
chezmoi version v2.72.0-SNAPSHOT-50b32b54c, commit 50b32b54ca5479935b53b408afa6b49d95996371, built at 2026-08-27T04:24:26Z, built by goreleaser
```
```
$ chezmoi-50b32b54c execute-template '{{ mul "7" "191" }}'
<no value>
```

With the officially released, latest chezmoi binary (2.72.0) that still uses sprig, string arguments are accepted:
```
chezmoi --version
chezmoi version v2.72.0, built at 2026-08-03T08:20:49Z
```
```
chezmoi execute-template '{{ mul "7" "191" }}'
1337
```

The [sprig docs](https://masterminds.github.io/sprig/math.html) say that the math functions expect integers:
> The following math functions operate on `int64` values.

However, in practice, they accept strings and apparently do some implicit conversion.

I don't know if the sprigin compatibility layer is even supposed to support all these undocumented quirks. If not, or if I have misunderstood something else, please ignore this post.

-----

Edit: There seems to be implicit conversion when the last parameter of `mul` is string.
```
$ chezmoi-50b32b54c execute-template '{{ mul "7" "191" }}'
<no value>
$ chezmoi-50b32b54c execute-template '{{ mul "7" 191 }}'
<no value>
$ chezmoi-50b32b54c execute-template '{{ mul 7 "191" }}'
1337
$ chezmoi-50b32b54c execute-template '{{ mul 7 191 }}'
1337
```

```
$ chezmoi execute-template '{{ mul "7" "191" }}'
1337
$ chezmoi execute-template '{{ mul 7 "191" }}'
1337
$ chezmoi execute-template '{{ mul "7" 191 }}'
1337
$ chezmoi execute-template '{{ mul 7 191 }}'
1337
```

**Comment by twpayne** [comment]
Thanks for reporting the problem with `mul`. An initial investigation reveals that it looks like sprout and sprig behave quite differently here: sprig converts all arguments to an `int64` using [`github.com/spf13/cast.ToInt64`](https://pkg.go.dev/github.com/spf13/cast#ToInt64) (ignoring any error), whereas sprout expects all arguments to be `float64`s. The same applies to many other arithmetic template functions.

Sprigin clearly needs some non-trivial changes here, so I'll revert back to sprig for now.
